sklearn/decomposition/_base.py

```diff
@@ -37,13 +37,18 @@ def get_covariance(self):
         cov : array of shape=(n_features, n_features)
             Estimated covariance of data.
         """
+        xp, _ = get_namespace(self.components_)
+
         components_ = self.components_
         exp_var = self.explained_variance_
         if self.whiten:
-            components_ = components_ * np.sqrt(exp_var[:, np.newaxis])
-        exp_var_diff = np.maximum(exp_var - self.noise_variance_, 0.0)
-        cov = np.dot(components_.T * exp_var_diff, components_)
-        cov.flat[:: len(cov) + 1] += self.noise_variance_  # modify diag inplace
+            components_ = components_ * xp.sqrt(exp_var[:, np.newaxis])
+        exp_var_diff = xp.maximum(
+            exp_var - self.noise_variance_, xp.zeros_like(exp_var)
+        )
+        cov = (components_.T * exp_var_diff) @ components_
+        # TODO use views instead?
+        cov.reshape(-1)[:: len(cov) + 1] += self.noise_variance_  # modify diag inplace
```

Comment:
We can not use `cov.reshape` because it is not a part of the Array API spec.

The Array API spec for [reshape](https://data-apis.org/array-api/latest/API_specification/generated/array_api.reshape.html?highlight=reshape) forces a view when `copy=False`, but NumPy's Array API does not follow the spec yet: https://github.com/numpy/numpy/issues/23410.

In the short term, I recommend using a `for` loop over the diagonal when `xp` is not NumPy.

```python
def add_to_diagonal(array, value, xp):
    if _is_numpy_namespace(xp):
        array_np = np.asarray(array)
        array_np.flat[:: array.shape[0] + 1] += value
        return xp.asarray(array_np)
    else:
        for i in range(array.shape[0]):
            array[i, i] += value
```

Reply:
Can we put things like this ("code that contains `if numpy: ... else: ...`") into the `_array_api.py` module? Or at least try to place it there to reduce the number of places we have to go hunt for such code fragments?

There can be namespace specific code that we leave in the estimator but I think it is worth making an effort to try and collect as much as possible in one place.